File: app/src/main/java/ru/radiationx/anilibria/api/releases/ReleaseParser.java
package ru.radiationx.anilibria.api.releases;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import ru.radiationx.anilibria.api.Api;
import ru.radiationx.anilibria.api.Client;

/**
 * Created by radiationx on 31.10.17.
 */

public class ReleaseParser {

    private static ArrayList<ReleaseItem> parseItems(String responseText) throws Exception {
        ArrayList<ReleaseItem> resItems = new ArrayList<>();
        final JSONObject responseJson = new JSONObject(responseText);
        final JSONArray jsonItems = responseJson.getJSONArray("items");
        for (int i = 0; i < jsonItems.length(); i++) {
            ReleaseItem item = new ReleaseItem();
            JSONObject jsonItem = jsonItems.getJSONObject(i);
            item.setId(jsonItem.getInt("id"));
            {
                String title = jsonItem.getString("title");
                String[] titles = title.split(" / ");
                for (int j = 0; j < titles.length; j++) {
                    title = titles[j];
                }
                item.setTitle(title);
            }

            item.setTorrentLink(jsonItem.getString("torrent_link"));
            item.setLink(jsonItem.getString("link"));
            item.setImage(jsonItem.getString("image"));
            item.setEpisodes(jsonItem.getString("episode"));
            item.setDescription(jsonItem.getString("description"));

            JSONArray jsonSeasons = jsonItem.getJSONArray("season");
            for (int j = 0; j < jsonSeasons.length(); j++) {
                item.addSeason(jsonSeasons.getString(j));
            }

            JSONArray jsonVoices = jsonItem.getJSONArray("season");
            for (int j = 0; j < jsonVoices.length(); j++) {
                item.addVoice(jsonVoices.getString(j));
            }

            JSONArray jsonGenres = jsonItem.getJSONArray("season");
            for (int j = 0; j < jsonGenres.length(); j++) {
                item.addGenre(jsonGenres.getString(j));
            }

            JSONArray jsonTypes = jsonItem.getJSONArray("season");
            for (int j = 0; j < jsonTypes.length(); j++) {
                item.addType(jsonTypes.getString(j));
            }

            resItems.add(item);
        }
        return resItems;
    }

    public static void releaseItemsAsync(int page, final Api.ApiCallback<ArrayList<ReleaseItem>> callback) {
        String url = "http://www.anilibria.tv/api/api.php?PAGEN_1=" + page;
        Client.get().get(url, new Client.ClientCallback() {
            @Override
            public void onError(Exception e) {
                callback.onError(e);
            }

            @Override
            public void onResponse(String responseText) throws Exception {
                ArrayList<ReleaseItem> resItems = parseItems(responseText);
                callback.onResponse(resItems);
            }
        });
    }
}

File: app/src/main/java/ru/radiationx/anilibria/model/data/remote/parsers/ReleaseParser.kt
package ru.radiationx.anilibria.model.data.remote.parsers

import org.json.JSONObject
import ru.radiationx.anilibria.entity.app.Paginated
import ru.radiationx.anilibria.entity.app.release.ReleaseFull
import ru.radiationx.anilibria.entity.app.release.ReleaseItem
import ru.radiationx.anilibria.entity.app.release.TorrentItem
import ru.radiationx.anilibria.extension.nullGet
import ru.radiationx.anilibria.extension.nullString
import ru.radiationx.anilibria.model.data.remote.Api
import ru.radiationx.anilibria.model.data.remote.IApiUtils
import javax.inject.Inject

/**
 * Created by radiationx on 18.12.17.
 */
class ReleaseParser @Inject constructor(
        private val apiUtils: IApiUtils
) {

    private fun parseRelease(jsonItem: JSONObject): ReleaseItem {
        val item = ReleaseItem()
        item.id = jsonItem.getInt("id")
        item.code = jsonItem.getString("code")
        item.names.addAll(jsonItem.getJSONArray("names").let { names ->
            (0 until names.length()).map {
                apiUtils.escapeHtml(names.getString(it)).toString()
            }
        })
        item.series = jsonItem.nullString("series")
        item.poster = Api.BASE_URL_IMAGES + jsonItem.nullString("poster")
        jsonItem.optJSONObject("favorite")?.also { jsonFavorite ->
            item.favoriteInfo.also {
                it.rating = jsonFavorite.getInt("rating")
                it.isAdded = jsonFavorite.getBoolean("added")
            }
        }
        item.torrentUpdate = jsonItem.nullString("last")?.let {
            try {
                it.toInt()
            } catch (ex: Exception) {
                item.torrentUpdate
            }
        } ?: item.torrentUpdate
        item.status = jsonItem.nullString("status")
        item.description = jsonItem.nullString("description")?.trim()

        jsonItem.nullString("type")?.also {
            item.types.add(it)
        }

        jsonItem.optJSONArray("genres")?.also { jsonGenres ->
            for (j in 0 until jsonGenres.length()) {
                item.genres.add(jsonGenres.getString(j))
            }
        }

        jsonItem.optJSONArray("voices")?.also { jsonVoices ->
            for (j in 0 until jsonVoices.length()) {
                item.voices.add(jsonVoices.getString(j))
            }
        }

        jsonItem.nullString("year")?.also {
            item.seasons.add(it)
        }

        jsonItem.nullString("day")?.also {
            item.days.add(it)
        }
        return item
    }

    fun releases(jsonResponse: JSONObject): Paginated<List<ReleaseItem>> {
        val resItems = mutableListOf<ReleaseItem>()
        val jsonItems = jsonResponse.getJSONArray("items")
        for (i in 0 until jsonItems.length()) {
            val jsonItem = jsonItems.getJSONObject(i)
            resItems.add(parseRelease(jsonItem))
        }
        val pagination = Paginated(resItems)
        val jsonNav = jsonResponse.getJSONObject("pagination")
        jsonNav.nullGet("page")?.let { pagination.page = it.toString().toInt() }
        jsonNav.nullGet("perPage")?.let { pagination.perPage = it.toString().toInt() }
        jsonNav.nullGet("allPages")?.let { pagination.allPages = it.toString().toInt() }
        jsonNav.nullGet("allItems")?.let { pagination.allItems = it.toString().toInt() }
        return pagination
    }

    fun release(jsonResponse: JSONObject): ReleaseFull {
        val baseRelease = parseRelease(jsonResponse)
        val release = ReleaseFull(baseRelease)

        jsonResponse.optJSONObject("blockedInfo")?.also {jsonBlockedInfo->
            release.blockedInfo.also {
                it.isBlocked = jsonBlockedInfo.getBoolean("blocked")
                it.reason = jsonBlockedInfo.nullString("reason")
            }
        }

        release.moonwalkLink = jsonResponse.nullString("moon")

        jsonResponse.optJSONArray("playlist")?.also { jsonPlaylist ->
            for (j in 0 until jsonPlaylist.length()) {
                val jsonEpisode = jsonPlaylist.getJSONObject(j)

                val episodeId: Int = jsonEpisode.optInt("id")
                val episodeTitle = jsonEpisode.nullString("title")

                ReleaseFull.Episode().also {
                    it.releaseId = release.id
                    it.id = episodeId
                    it.title = episodeTitle
                    it.urlSd = jsonEpisode.nullString("sd")
                    it.urlHd = jsonEpisode.nullString("hd")
                    it.type = ReleaseFull.Episode.Type.ONLINE
                    release.episodes.add(it)
                }

                ReleaseFull.Episode().also {
                    it.releaseId = release.id
                    it.id = episodeId
                    it.title = episodeTitle
                    it.urlSd = jsonEpisode.nullString("srcSd")
                    it.urlHd = jsonEpisode.nullString("srcHd")
                    it.type = ReleaseFull.Episode.Type.SOURCE
                    release.episodesSource.add(it)
                }
            }
        }

        jsonResponse.getJSONArray("torrents")?.also { jsonTorrents ->
            for (j in 0 until jsonTorrents.length()) {
                jsonTorrents.optJSONObject(j)?.let { jsonTorrent ->
                    release.torrents.add(TorrentItem().apply {
                        id = jsonTorrent.optInt("id")
                        hash = jsonTorrent.nullString("hash")
                        leechers = jsonTorrent.optInt("leechers")
                        seeders = jsonTorrent.optInt("seeders")
                        completed = jsonTorrent.optInt("completed")
                        quality = jsonTorrent.nullString("quality")
                        series = jsonTorrent.nullString("series")
                        size = jsonTorrent.optLong("size")
                        url = "${Api.SITE_URL}${jsonTorrent.nullString("url")}"
                    })
                }
            }
        }

        release.showDonateDialog = jsonResponse.optBoolean("showDonateDialog")

        return release
    }


}
